### PR TITLE
Preview: xUnit-based Dafny test suite

### PR DESCRIPTION
**Pull request by robin-aws**
See [Test/README.md](Test/README.md) for details.

I've just refactored the majority of the comp/ directory. As indicated in the TODOs in the README file, my plan is to write a bit of a script to convert the rest of the test suite, which will reduce the risk of this refactoring merging badly with other PRs.

### CODE REVIEW

File: Test/DafnyTests/DafnyTests.cs [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+using System.Linq;
+using System.Reflection;
+using Xunit;
+using Xunit.Sdk;
+using YamlDotNet.RepresentationModel;
+using YamlDotNet.Serialization;
+
+namespace DafnyTests {
+
+    public class DafnyTests {
```
Comment (RustanLeino) [inline Test/DafnyTests/DafnyTests.cs:14]: Use 2-space indentation.

File: Test/comp/JsModule.dfy [hunk 2]
```diff
@@ -1,6 +1,8 @@
-// RUN: %dafny /compile:3 "%s" /compileTarget:js > "%t"
-// note: putting /compileTarget:js after "%s" overrides user-provided option
-// RUN: %diff "%s.expect" "%t"
+/*
+---
+compile: 3
+compileTarget: js
```
Comment (RustanLeino) [inline Test/comp/JsModule.dfy:4]: This seems like a nice and descriptive way to declare special parameters for those test cases that need it.

File: Test/DafnyTests/YamlTests/multiple.expect.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+- a: b
```
Comment (RustanLeino) [inline Test/DafnyTests/YamlTests/multiple.expect.yml:1]: What are the files in this YamlTests folder examples of?

**Comment by robin-aws** [comment]
> This new format looks clean and usable. I also like it that it avoids .expect files that for compilation duplicate the output once for each compiler (which is verbose and could be error prone).

Yes! And it's been very instructive finding all the places that duplicated output actually differed slightly. :)

> How do you add a new test? Just add the .dfy file and an .expect file and the framework does the rest, or do you also need to list the test in some configuration file?

Yup, it's just searching for all *.dfy files recursively in the Test directory. Unless explicitly configured with a "compile" setting, it will create a /compile:3 test case for each target language language. I still have to apply David's optimization of verifying separately.

File: Test/DafnyTests/DafnyTests.cs (hunk 1 above)
Comment (robin-aws) [reply Test/DafnyTests/DafnyTests.cs:14]: Can do. The implementation in general needs cleaning up before the non-draft PR.

File: Test/comp/JsModule.dfy (hunk 2 above)
Comment (robin-aws) [reply Test/comp/JsModule.dfy:4]: Thanks, I agree! I really didn't want to invent anything new that had to parse myself. There are lots of examples of using JSON for data-driven tests but I find YAML cleaner and easier to read, and it's a superset of JSON anyway for more complicated cases.

**Comment by davidcok** [comment]
A significant number of tests just check for parse or solution errors — so just /compile:0 for those

File: Test/DafnyTests/YamlTests/multiple.expect.yml (hunk 3 above)
Comment (robin-aws) [reply Test/DafnyTests/YamlTests/multiple.expect.yml:1]: This is read by the `ExpandTest` test, a meta-test for the logic to convert list values in the configuration to lists of configurations. It probably belongs in the XUnitExtensions namespace instead.

**Comment by robin-aws** [comment]
> A significant number of tests just check for parse or solution errors — so just /compile:0 for those

Indeed, when I refactor the rest of the tests a lot of them will have the "compile: 0" configuration. The default of compiling and running for every target language is at least safe if slow. Rustan and I also discussed omitting the *.expect file for test files that are expected to verify successfully with no additional warning or info messages, but I will probably leave that as a follow-up improvement.

I considered supporting implicit shared config based on which folder a test sits in, but for at least the first refactoring I'd rather not move hundreds of files around.

**Comment by parno** [comment]
I agree that this looks like a cleaner and more sustainable way of specifying and checking test cases.  Thanks for pushing it ahead!  

I was a bit puzzled by this comment in the `README`: 

> Such exceptions are automatically flagged as "known issues" and classified as a "skipped/ignored" test.

If there's a custom `.expect` file for a specific language, and the output matches those expectations, shouldn't it be a successful test?  Maybe I'm misunderstanding the comment.

**Comment by robin-aws** [comment]
> I was a bit puzzled by this comment in the `README`:
> 
> > Such exceptions are automatically flagged as "known issues" and classified as a "skipped/ignored" test.
> 
> If there's a custom `.expect` file for a specific language, and the output matches those expectations, shouldn't it be a successful test? Maybe I'm misunderstanding the comment.

I'm implicitly making the strong statement that any inconsistency in output between the target languages is a bug. The idea is to assert the output is as expected, but to mark the test as a known issue that should be fixed eventually. It will become a FAIL if there is any change though, such as the inconsistency getting addressed, at which point the exception can be removed.

Note that if a test specifies `compileTarget` then that's treated as a valid language-specific test case.

**Comment by davidcok** [comment]
Note that there is also the case of the test intentionally returning a non-zero exit value. With lit, such cases are automatic failures so one can’t test, for example, erroneous command lines.

In general, along with checking expected standard out, one should have an expected exit code and expected stderr.

**Comment by robin-aws** [comment]
> Note that there is also the case of the test intentionally returning a non-zero exit value. With lit, such cases are automatic failures so one can’t test, for example, erroneous command lines. In general, along with checking expected standard out, one should have an expected exit code and expected stderr.

Indeed, and part of converting the rest of the tests will include adding some more options that don't directly map to dafny CLI switches. I also still have to support partial matches on stdout/stderr for cases where the output isn't 100% context-free.

File: Source/LitTestConverter/LitTestConvertor.cs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        verifyOnlyCount++;
+        
+      } else if (testConfigs.Count(c => c.CompileTarget != null) > 1) {
+        defaultCount++;
+      }
+
+      IEnumerable<string> otherLines = lines.Skip(litCommands.Count);
+      
+
+    }
```
Comment (acioc) [inline Source/LitTestConverter/LitTestConvertor.cs:70]:
```suggestion
    }
```

File: Source/LitTestConverter/LitTestConvertor.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+    private static bool DictionaryContainsEntry<K, V>(Dictionary<K, V> dictionary, K key, V value) {
```
Comment (acioc) [inline Source/LitTestConverter/LitTestConvertor.cs:72]: Might be worth validating these are all non-null as required.

File: Source/LitTestConverter/LitTestConvertor.cs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        } else {
+          return new KeyValuePair<string, string>(argument, "yes");
+        }
+      } else {
+        return new KeyValuePair<string, string>(TEST_CONFIG_OTHER_FILES, argument);
+      }
+    }
+
+    public void Run(string root) {
+      // TODO-RS: Search for "*.transcript" too
```
Comment (acioc) [inline Source/LitTestConverter/LitTestConvertor.cs:153]: Flagging TODO

File: Source/LitTestConverter/LitTestConvertor.cs [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+      if (!arguments[^1].Equals("\"%t\"") 
+          || !(arguments[^2].Equals(">") || arguments[^2].Equals(">>")))
+      {
+        throw new ArgumentException("Non-standard output in %dafny command: " + dafnyCommand);
+      }
+      arguments.RemoveRange(arguments.Count - 2, 2);
+
+      if (!arguments.Remove("\"%s\"")) {
+        spec.DafnyArguments[TEST_CONFIG_INCLUDE_THIS_FILE] = "no";
+      }
```
Comment (acioc) [inline Source/LitTestConverter/LitTestConvertor.cs:115]: Would be worth adding a comment here for why we want to do this.

File: Source/DafnyTests/DafnyTestCase.cs [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+using YamlDotNet.Serialization.NamingConventions;
+using YamlDotNet.Serialization.ObjectFactories;
+
+namespace DafnyTests {
+  
+  public class DafnyTestSpec : IEnumerable<DafnyTestCase> {
+
+    private static DirectoryInfo OUTPUT_ROOT = new DirectoryInfo(Directory.GetCurrentDirectory());
+    
+    // TODO-RS: This is an ugly method of locating the project root.
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:24]: Flagging TODO (I flag TODO's to make sure they're either addressed or we have a separate issue to fix them in the future).

File: Source/DafnyTests/DafnyTestCase.cs [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+    // The proper fix is to run entirely out of the output directory,
+    // and the projects are at least partially configured to make that possible,
+    // but it's not quite working yet.
+    private static string DAFNY_ROOT = 
+      OUTPUT_ROOT.Parent.Parent.Parent.Parent.Parent.FullName;
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:29]: :(

File: Source/DafnyTests/DafnyTestCase.cs [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+
+    public static readonly string TEST_ROOT = Path.Combine(DAFNY_ROOT, "Test") + Path.DirectorySeparatorChar;
+    public static readonly string COMP_DIR = Path.Combine(TEST_ROOT, "comp") + Path.DirectorySeparatorChar;
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:32]: np: I wish all of these folders followed similar capitalization. Def don't expect this to be fixed.

File: Source/DafnyTests/DafnyTestCase.cs [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+    
+    public Dictionary<string, object> DafnyArguments = new Dictionary<string, object>();
+    public List<string> OtherFiles = new List<string>();
+    
+    public Dictionary<string, DafnyTestSpec> CompileTargetOverrides = new Dictionary<string, DafnyTestSpec>();
+    
+    public DafnyTestCase.Expectation Expected;
+
+    public DafnyTestSpec(string sourcePath) {
+      SourcePath = sourcePath;
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:54]: Should we ensure this string is non-empty?

File: Source/DafnyTests/DafnyTestCase.cs [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+      return ResolveCompile()
+        .SelectMany(spec => spec.ExpandArguments())
+        .Select(spec => spec.ResolveExpected().ToTestCase())
+        .GetEnumerator();
+    }
+
+    public string Compile {
+      get {
+        if (DafnyArguments.TryGetValue(DAFNY_COMPILE_OPTION, out var compile)) {
+          return (string) compile;
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:74]: np
```suggestion
          return (string)compile;
```

File: Source/DafnyTests/DafnyTestCase.cs [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+        return null;
+      }
+      set {
+        DafnyArguments[DAFNY_COMPILE_OPTION] = value;
+      }
+    }
+    public string CompileTarget {
+      get {
+        if (DafnyArguments.TryGetValue(DAFNY_COMPILE_TARGET_OPTION, out var compileTarget)) {
+          return (string) compileTarget;
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:85]: np
```suggestion
          return (string)compileTarget;
```

File: Source/DafnyTests/DafnyTestCase.cs [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+      }
+    }
+
+    private IEnumerable<DafnyTestSpec> ResolveCompile() {
+      if (Compile == null) {
+        Compile = "3";
+      }
+      
+      if ("3".Equals(Compile) && CompileTarget == null) {
+        var compileTargets = new[] {"cs", "java", "go", "js"};
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:97]: We should probably have a static const of all supported languages somewhere higher so we can reduce the chance of missing this as new compile targets are added.

File: Source/DafnyTests/DafnyTestCase.cs [hunk 15]
```diff
@@ -0,0 +1,417 @@
...
+
+    private static IEnumerable<KeyValuePair<string, object>> ExpandValue(KeyValuePair<string, object> pair) {
+      return Expand(pair.Value).Select(v => new KeyValuePair<string, object>(pair.Key, v));
+    }
+  }
+
+  public class ForEachArgumentList : List<string> { }
+  
+  // TODO-RS: This is really close to a generic CLI test case -
+  // can we move all the dafny-specific stuff to DafnyTestSpec instead?
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:181]: Flagging TODO. Also, I would support this.

File: Source/DafnyTests/DafnyTestCase.cs [hunk 16]
```diff
@@ -0,0 +1,417 @@
...
+    private IEnumerable<DafnyTestSpec> ExpandArguments() {
+      return DafnyArguments.Select(ExpandValue)
+                           .CartesianProduct()
+                           .Select(e => e.ToDictionary(p => p.Key, p => p.Value))
+                           .Select(args => new DafnyTestSpec(SourcePath, args, OtherFiles, Expected));
+    }
+
+    public static IEnumerable<string> Expand(object obj) {
+      if (obj is ForEachArgumentList forEach) {
+        return forEach;
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:167]: np because it takes a second to realize that `forEach` represents the list and not an argument within the list. I know we use `foreach` in other contexts, but I think this would make it a bit clearer here, at least.

```suggestion
      if (obj is ForEachArgumentList forEachArgumentList) {
        return forEachArgumentList;
```

File: Source/DafnyTests/DafnyTestCase.cs [hunk 17]
```diff
@@ -0,0 +1,417 @@
...
+
+        // We do not want output such as "Compiled program written to Foo.cs"
+        // from the compilers, since that changes with the target language
+        "/compileVerbose:0"
+      };
+      dafnyArguments.AddRange(arguments);
+
+      using (Process dafnyProcess = new Process()) {
+        dafnyProcess.StartInfo.FileName = DafnyTestSpec.DAFNY_EXE;
+        foreach(var argument in dafnyArguments) {
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:265]: np to match other uses of `foreach` in this file (also just generally)
```suggestion
        foreach (var argument in dafnyArguments) {
 ```

File: Source/DafnyTests/DafnyTestCase.cs [hunk 18]
```diff
@@ -0,0 +1,417 @@
...
+      info.AddValue(nameof(Expected), Expected);
+    }
+    
+    public void Deserialize(IXunitSerializationInfo info) {
+      Arguments = info.GetValue<string[]>(nameof(Arguments));
+      Expected = info.GetValue<Expectation>(nameof(Expected));
+    }
+
+    public static ProcessResult RunDafny(IEnumerable<string> arguments) {
+      // TODO-RS: Let these be overridden
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:248]: Flagging TODO

File: Source/DafnyTests/DafnyTestCase.cs [hunk 19]
```diff
@@ -0,0 +1,417 @@
...
+      List<string> dafnyArguments = new List<string> {
+        // Expected output does not contain logo
+        "/nologo",
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:251]: I don't expect this to be changed, but `nologo` actually refers to the copyright info and version, right (not actually a logo)? So this should really be renamed to `/noDafnyCopyrightInfo` or something more like that?

File: Source/DafnyTests/DafnyTestCase.cs [hunk 20]
```diff
@@ -0,0 +1,417 @@
...
+    }
+  }
+
+  
+  public class DafnyTestYamlDataDiscoverer : YamlDataDiscoverer {
+
+    private const string DEFAULT_CONFIG = 
+@"!dafnyTestSpec
+dafnyArguments: {}
+";
```
Comment (acioc) [inline Source/DafnyTests/DafnyTestCase.cs:331]: np: Can we format this better?

**Review by acioc — COMMENTED** [review]
Overall looks good. It looks like GitHub now allows you to upload mp4 in comments, so it would be great to see this in action in an IDE (like you mentioned in the PR).

**Comment by keyboardDrummer** [comment]
I'm exciting about this PR. It's cool that you can dynamically discover the XUnit tests like that.

The new format for tests also seems interesting, although from the large amount of changed files I had some trouble discovering what the new format exactly is. 

Would it be an option to scope down this PR to only replace `lit` with `XUnit`, but not change the test format? You mention "my plan is to write a bit of a script to convert the rest of the test suite, which will reduce the risk of this refactoring merging badly with other PRs.", I'm guessing this script needs to understand the lit format, so we might as well use it to support the lit format, then we can convert the tests to a new format in a separate CR.

**Comment by robin-aws** [comment]
Thanks @keyboardDrummer! 

My intention was actually to just aim to add the new test format support in one PR, with perhaps a handful of existing tests converted over, but without removing the lit support or touching the whole regression suite yet. Then the rest could be converted in batches to reduce the chances of bad conflicts with other PRs. 

You're right that we could just convert to xUnit while keeping the lit syntax, but I think that would still need a fair bit more work to understand 100% of the huge test suite. I also worry it would be confusing for contributors to have lit tests you don't run with lit, especially if not all of lit was supported.

There's a bit of description of the new test format in `Test/README.md` although it's obviously not completed yet. I'm not completely satisfied with it yet and we should review it separately before merging anything.

**Comment by robin-aws** [comment]
One other thought: `DafnyLanguageServer.Test` was added to this repository since I initially opened this PR, which has quite a few NUnit tests. It made me really happy to see Dafny finally get some "unit tests" even if I wasn't the one to add the first ones. :) It would make sense to convert those to xUnit as well though - AFAICT there's nothing NUnit-specific in there and they should be easy to switch over.

File: Test/README.md [hunk 21]
```diff
@@ -0,0 +1,20 @@
+# Tests
+
+Dafny test cases are based on single Dafny source files together with the expected output from the `dafny` CLI tool. If the source file is named `Foo.dfy`, the expected output will contained in the file named `Foo.dfy.expect`.
+
+The default behaviour is to assert that the source file verifies successfully and, for each currently-supported target language, can be compiled and run to produce the expected output. When the behaviour is not 100% consistent between different target languages, expected discrepancies can be recorded in additional files with the name `Foo.dfy.<language>.expect`. For example: [Test/comp/NativeNumbers.dfy.js.expect](Test/comp/NativeNumbers.dfy.js.expect). Such exceptions are automatically flagged as "known issues" and classified as a "skipped/ignored" test.
```
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Test/README.md:5]: > The default behaviour is to assert that the source file verifies successfully and, for each currently-supported target language, can be compiled and run to produce the expected output.

Hmm, and if the program fails to verify, is the default behavior then not to compile it and compare the verification errors with the expected output?

I also wonder whether we can run the test suite in a 'compilation mode', where all verification is skipped, or a verification mode, where all compilation is skipped.

File: Source/DafnyTests/TemporaryDirectory.cs [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace DafnyTests {
+  public class TemporaryDirectory : IDisposable {
+    public readonly DirectoryInfo DirInfo;
+
+    public TemporaryDirectory(string parent, string prefix = "") {
+      string dirPath;
+      do {
+        dirPath = Path.Combine(parent, prefix + Path.GetRandomFileName());
+      } while (File.Exists(dirPath) || Directory.Exists(dirPath));
```
[the remark this answers is not in the document]
Comment (robin-aws) [reply Source/DafnyTests/TemporaryDirectory.cs:12]: It loops because if we're really unlucky we could get a name collision more than once. Adding a comment.

File: Source/DafnyTests/DafnyTests.csproj [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+    <PropertyGroup>
+        <TargetFramework>netcoreapp3.1</TargetFramework>
+        <AssemblyName>DafnyTests</AssemblyName>
+        <GenerateEmbeddedFilesManifest>true</GenerateEmbeddedFilesManifest>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="DiffPlex" Version="1.6.3" />
+        <PackageReference Include="FluentAssertions" Version="5.10.3" />
+        <PackageReference Include="Microsoft.Extensions.FileProviders.Embedded" Version="5.0.0-rc.1.20451.17" />
```
[the remark this answers is not in the document]
Comment (robin-aws) [reply Source/DafnyTests/DafnyTests.csproj:12]: Good callout, I don't know why Rider picked a release candidate as the latest version. Upgraded to 5.0.8 now (since it's been over a year since I touched this PR :)

File: Source/DafnyTests/XUnitExtensions/CollectionPerTestCaseTheoryDiscoverer.cs [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public IEnumerable<IXunitTestCase> Discover(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo factAttribute) {
+      // This discoverer requires pre-enumeration in order to assign a collection to each test case.
+      discoveryOptions.SetValue("xunit.discovery.PreEnumerateTheories", true);
+            
+      IEnumerable<IXunitTestCase> testCases = theoryDiscoverer.Discover(discoveryOptions, testMethod, factAttribute);
+      var shardEnvVar = Environment.GetEnvironmentVariable("XUNIT_SHARD");
+      var numShardsEnvVar = Environment.GetEnvironmentVariable("XUNIT_SHARD_COUNT");
+      // TODO-RS: More careful error checking
```
[the remark this answers is not in the document]
Comment (robin-aws) [reply Source/DafnyTests/XUnitExtensions/CollectionPerTestCaseTheoryDiscoverer.cs:30]: Yup, and `shard <= numShards`. Fixed.

File: Test/README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (robin-aws) [reply Test/README.md:5]: No, the default is to expect the file is valid and can be executed. I wouldn't want to automatically fall back to assuming the file is intentionally invalid and meant to assert what the verification errors are, on the off chance that could produce a false positive. But all it takes to mark a file as a "expected validation failures" test case is adding `compile: 0`.

Being able to filter the cases and/or alter the parameters when running the tests is definitely something I want to support, but only after we get all the existing test cases converted over.

File: Source/DafnyTests/DafnyTestCase.cs (hunk 10 above)
Comment (robin-aws) [reply Source/DafnyTests/DafnyTestCase.cs]: Not with that attitude :) Definitely not in THIS PR though!

File: Source/DafnyTests/DafnyTestCase.cs (hunk 14 above)
Comment (robin-aws) [reply Source/DafnyTests/DafnyTestCase.cs]: Couldn't agree more, this should just refer directly to the `CompilationTarget` enumeration. I've only got it as a separate list temporarily and already added a TODO here.

**Comment by robin-aws** [comment]
FYI to all interested parties: my plan here is to revert all the changes to convert existing tests for now, and to just add a bunch of sample test files (old and new format) to use as unit tests for both the new test runner and the LitTestConvertor utility. That will help document the new format as well. 

Then we can start converting existing tests using the convertor in batches over multiple PRs, to keep them simpler and easier to review, and reduce the changes of conflicting badly with other PRs (although again having the convertor should greatly reduce the pain of merge conflicts).

File: docs/DafnyRef/examples/Example-TwoState.dfy [hunk 25]
```diff
@@ -0,0 +1,18 @@
+
+// This file only exists to include these examples as a test case.
```
Comment (keyboardDrummer) [inline docs/DafnyRef/examples/Example-TwoState.dfy:2]: Do you know why is this here as a test-case but not used in the reference manual?

File: Test/git-issues/git-issue-801.dfy [hunk 26]
```diff
@@ -1,6 +0,0 @@
-// RUN: %dafny /compile:0 /verifyAllModules "%s" > "%t"
```
Comment (keyboardDrummer) [inline Test/git-issues/git-issue-801.dfy:1]: Nice cleanup :)

File: Test/refman/Example-Fail1.dfy [hunk 27]
```diff
@@ -1,4 +1,2 @@
-// RUN: %dafny /verifyAllModules /compile:0 "%s" > "%t"
```
Comment (keyboardDrummer) [inline Test/refman/Example-Fail1.dfy:1]: Nice change :)

File: Source/IntegrationTests/LitTests.cs [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+
+    private readonly ITestOutputHelper output;
+
+    public LitTests(ITestOutputHelper output) {
+      this.output = output;
+    }
+
+    [FileTheory]
+    [FileData(Includes = new[] { "**/*.dfy", "**/*.transcript" },
+              Excludes = new[] { "**/Inputs/**/*", "refman/examples/**/*" })]
```
Comment (keyboardDrummer) [inline Source/IntegrationTests/LitTests.cs:93]: I had a few output files from a previous run that were accidentally picked up as Lit tests, for example this one:

`TestFiles/LitTests/LitTest/allocated1/dafny0/Output/Calculations.dfy.tmp.dprint.dfy`

**Review by keyboardDrummer — COMMENTED** [review]
Whoa, this was not an easy change!

Looks great. I tried the in-IDE runner and it works well.

I also tried to trigger the new test path through the CLI with `dotnet test -v:n Source/IntegrationTests/IntegrationTests.csproj`

It built the solution as expected, but then halted after showing
`
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
`

I believe it was running all the tests under the hood, but without providing any feedback that wasn't useful.

I think that before we can remove `lit`, we should have an equally good CLI test experience.

File: .github/workflows/msbuild.yml [hunk 29]
```diff
@@ -47,6 +47,8 @@ jobs:
         repository: boogie-org/boogie
         path: boogie
         ref: v${{ steps.regex-match.outputs.group1 }}
+    - name: Build Java runtime
```
[the remark this answers is not in the document]
Comment (robin-aws) [reply .github/workflows/msbuild.yml:50]: YES, I would love that. Worth a time-boxed attempt to get that working if it would mean avoiding adding this to basically every build job, no to mention simplifying the contributor build process.

Note that I'm implicitly dropping the lit behavior in some sub-directories that marks them unsupported if certain language runtimes aren't available. I'm intentionally simplifying things to say that those are hard dependencies and the tests are always run. That makes it reasonable for the `DafnyRuntime` project to have a hard dependency on `javac` in my opinion.

**Comment by robin-aws** [comment]
> Whoa, this was not an easy change! Looks great. I tried the in-IDE runner and it works well.

I've never felt more agreement with the "chore:" conventional commit label. :) But I think it's worth it and an important first step towards a better state!

> I also tried to trigger the new test path through the CLI with `dotnet test -v:n Source/IntegrationTests/IntegrationTests.csproj`
> 
> It built the solution as expected, but then halted after showing `Starting test execution, please wait... A total of 1 test files matched the specified pattern.`
> 
> I believe it was running all the tests under the hood, but without providing any feedback that wasn't useful. I think we should have an equally good CLI test experience before we can remove `lit`.

That's odd, I've been using the `dotnet test` CLI lots myself while developing it, and I get output quite quickly when I run `dotnet test -v:n Source/IntegrationTests/IntegrationTests.csproj`:

```
...
     1>BuildProject:
         Build completed.
         
Test run for /Users/salkeldr/Documents/GitHub/dafny/Source/IntegrationTests/bin/Debug/net5.0/IntegrationTests.dll (.NETCoreApp,Version=v5.0)
Microsoft (R) Test Execution Command Line Tool Version 16.10.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.3+1b45f5407b (64-bit .NET 5.0.8)
[xUnit.net 00:00:00.91]   Discovering: IntegrationTests
[xUnit.net 00:00:01.12]   Discovered:  IntegrationTests
[xUnit.net 00:00:01.12]   Starting:    IntegrationTests
  Passed allocated1/Allocated2.dfy [1 s]
  Passed allocated1/dafny0/AssertBy.dfy [3 s]
  Passed allocated1/dafny0/AssumptionVariables1.dfy [3 s]
  Passed allocated1/dafny0/AdvancedLHS.dfy [4 s]
  Passed allocated1/dafny0/BadFunction.dfy [4 s]
  Passed allocated1/dafny0/Backticks.dfy [7 s]
  Passed allocated1/dafny0/BindingGuards.dfy [6 s]
  Passed allocated1/Allocated1.dfy [13 s]
...
```

I wonder if it's not picking up the `-v:n` argument to increase logging verbosity for some reason?

I was toying with providing a tiny `xlit` script that would imitate a little bit of the `lit` CLI, so you can still just say `xlit <file>` or `xlit <directory>`.

File: Source/IntegrationTests/LitTests.cs (hunk 28 above)
Comment (robin-aws) [reply Source/IntegrationTests/LitTests.cs]: Right, lit automatically filters out Output files internally. I've addressed this by adding another `Exclude` here even though the lit test runner should really be responsible for that, since it's a fair bit easier.

Ideally it would be automatic, but another improvement I want to make after this PR is to run each test in a fresh temporary directory instead of a fixed and shared `Output` directory (I started doing that in earlier versions of this PR in fact but have since trimmed that scope for v1), which will avoid potentially polluting results and also make this issue moot.

File: docs/DafnyRef/examples/Example-TwoState.dfy (hunk 25 above)
Comment (robin-aws) [reply docs/DafnyRef/examples/Example-TwoState.dfy:2]: Yup, it's just because the section that uses these examples (https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-two-state) doesn't ever actually define the `Cell` class, but it has to be defined here in order for the examples to resolve as complete Dafny programs.

File: .github/workflows/xunit-tests.yml [hunk 30]
```diff
@@ -48,7 +48,7 @@ jobs:
       shell: pwsh
       run: |
         Invoke-WebRequest ${{env.z3BaseUri}}/${{matrix.z3}}.zip -OutFile z3.zip
-        Expand-Archive z3.zip Binaries/z3
+        Expand-Archive z3.zip ./Binaries/z3
```
[the remark this answers is not in the document]
Comment (robin-aws) [reply .github/workflows/xunit-tests.yml]: Thanks for the suggestion @MikaelMayer! I did eventually get it working in the later commits, but there's probably still a better approach so let me know if I can improve it.

File: Source/IntegrationTests/IntegrationTests.csproj [hunk 31]
```diff
@@ -0,0 +1,45 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <TargetFramework>net5.0</TargetFramework>
+        <GenerateEmbeddedFilesManifest>true</GenerateEmbeddedFilesManifest>
+        <IsPackable>false</IsPackable>
+        <Nullable>enable</Nullable>
+        <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
Comment (keyboardDrummer) [inline Source/IntegrationTests/IntegrationTests.csproj:8]: Please remove this for now since it collides with the 'singletons' test run that regretfully currently runs StyleCop because it accidentally picks that up from Boogie.

As long a local `dotnet build Dafny/Source.sln` builds without warnings, I think we're good to merge this PR.

**Review by keyboardDrummer — APPROVED** [review]

**Review by keyboardDrummer — APPROVED** [review]
